Avoid littering postgres server logs with "could not obtain lock" with HA schedulers

ashb opened the pull request:
If you are running multiple schedulers on PostgreSQL, it is likely that sooner or later you will have one scheduler fail the race to enter the critical section (which is fine, and expected).

However this can end up spamming the DB logs with errors like this:

```
Nov 26 14:08:48 sinope postgres[709953]: 2021-11-26 14:08:48.672 GMT [709953] ERROR:  could not obtain lock on row in relation "slot_pool"
Nov 26 14:08:48 sinope postgres[709953]: 2021-11-26 14:08:48.672 GMT [709953] STATEMENT:  SELECT slot_pool.pool AS slot_pool_pool, slot_pool.slots AS slot_pool_slots
Nov 26 14:08:48 sinope postgres[709953]:         FROM slot_pool FOR UPDATE NOWAIT
Nov 26 14:08:49 sinope postgres[709954]: 2021-11-26 14:08:49.730 GMT [709954] ERROR:  could not obtain lock on row in relation "slot_pool"
Nov 26 14:08:49 sinope postgres[709954]: 2021-11-26 14:08:49.730 GMT [709954] STATEMENT:  SELECT slot_pool.pool AS slot_pool_pool, slot_pool.slots AS slot_pool_slots
Nov 26 14:08:49 sinope postgres[709954]:         FROM slot_pool FOR UPDATE NOWAIT
```

If you are really unlucky that can end up happening over and over and over again.

So to avoid this error, for PostgreSQL only, we first try to acquire an "advisory lock" (advisory because it's up to the application to respect it), and if we cannot raise an error _like_ would have happened from the `FOR UPDATE NOWAIT`.

(We still obtain the exclusive log on the pool rows so that the rows are locked.)

This PR is split in to two commits, the second obtains the lock, and the first commit refactors the existing global locks to use enums to remove magic constants as these (integer for postgres) lock ids are "global", so we need to be sure the scheduler's lock doesn't clash with the `db upgrade` lock.

ashb committed:
Remove magic constants from global DB locks

And `create_global_lock` isn't anything to do with session, so I have
moved it to utils.db instead.

And as part of this I changed the lock id that `airflow db reset` uses
to share one with `airflow db upgrade` -- there's no point blocking
upgrade if a reset is going to clobber it halfway through.

ashb committed:
Avoid littering Postgres server logs with "could not obtain lock"

If you are running multiple schedulers on PostgreSQL, it is likely that
sooner or later you will have one scheduler fail the race to enter the
critical section (which is fine, and expected).

However this can end up spamming the DB logs with errors like this:

```
Nov 26 14:08:48 sinope postgres[709953]: 2021-11-26 14:08:48.672 GMT [709953] ERROR:  could not obtain lock on row in relation "slot_pool"
Nov 26 14:08:48 sinope postgres[709953]: 2021-11-26 14:08:48.672 GMT [709953] STATEMENT:  SELECT slot_pool.pool AS slot_pool_pool, slot_pool.slots AS slot_pool_slots
Nov 26 14:08:48 sinope postgres[709953]:         FROM slot_pool FOR UPDATE NOWAIT
Nov 26 14:08:49 sinope postgres[709954]: 2021-11-26 14:08:49.730 GMT [709954] ERROR:  could not obtain lock on row in relation "slot_pool"
Nov 26 14:08:49 sinope postgres[709954]: 2021-11-26 14:08:49.730 GMT [709954] STATEMENT:  SELECT slot_pool.pool AS slot_pool_pool, slot_pool.slots AS slot_pool_slots
Nov 26 14:08:49 sinope postgres[709954]:         FROM slot_pool FOR UPDATE NOWAIT
```

If you are really unlucky that can end up happening over and over and
over again.

So to avoid this error, for PostgreSQL only, we first try to acquire an
"advisory lock" (advisory because it's up to the application to respect
it), and if we cannot raise an error _like_ would have happened from the
`FOR UPDATE NOWAIT`.

(We still obtain the exclusive log on the pool rows so that the rows are locked.)

kaxil commented on airflow/utils/db.py:
```diff
@@ -986,3 +988,47 @@ def check(session=None):
+    try:
+        if dialect.name == 'postgresql':
+            conn.execute(text('SET LOCK_TIMEOUT to :timeout'), timeout=lock_timeout)
+            conn.execute(text('SELECT pg_advisory_lock(:id)'), id=lock.value)
```
```suggestion
            conn.execute(text('SELECT PG_ADVISORY_LOCK(:id)'), id=lock.value)
```

kaxil replied:
nit as others are CAP in above and below lines

ashb replied:
https://www.postgresql.org/docs/9.1/functions-admin.html -- casing as per the docs.

kaxil commented on airflow/utils/db.py:
```diff
@@ -986,3 +988,47 @@ def check(session=None):
+    finally:
+        if dialect.name == 'postgresql':
+            conn.execute('SET LOCK_TIMEOUT TO DEFAULT')
+            conn.execute(text('SELECT pg_advisory_unlock(:id)'), id=lock.value)
```
```suggestion
            conn.execute(text('SELECT PG_ADVISORY_UNLOCK(:id)'), id=lock.value)
```

kaxil commented on airflow/utils/db.py:
```diff
@@ -986,3 +988,47 @@ def check(session=None):
+    """
+
+    INIT = enum.auto()
+    MIGRATIONS = enum.auto()
```
Any reason we need both `init` and `migrations` as `initdb()` runs `upgradedb()` too?

ashb replied:
Yeah, I wasn't sure. Previously there were _three_ separate locks: one for upgrade, one for reset, and one for the "DAG sync" part of init.

This init lock could perhaps be dropped entirely, yes

ashb replied:
Fixed in f18e2cdf2

kaxil commented on airflow/jobs/scheduler_job.py:
```diff
@@ -234,8 +234,25 @@ def _executable_task_instances_to_queued(self, max_tis: int, session: Session =
+            # COMMIT/ROLLBACK)
+            lock_acquired = session.execute(
+                text("SELECT pg_try_advisory_xact_lock(:id)").bindparams(
+                    id=DBLocks.SCHEDULER_CRITICAL_SECTION
```
```suggestion
                    id=DBLocks.SCHEDULER_CRITICAL_SECTION.value
```

ashb replied:
Thanks -- it works anyway (as its an IntEnum) but better to be explicit I agree.

ashb committed:
Update airflow/jobs/scheduler_job.py

Co-authored-by: Kaxil Naik <kaxilnaik@gmail.com>

ashb committed:
fixup! Remove magic constants from global DB locks

ashb committed:
Use same db lock for initdb and upgradedb

They do conceptually very similar things and when one is running the
other shouldn't either. (This is unlikely to ever be hit in practice)

uranusjr commented on airflow/utils/db.py:
```diff
@@ -986,3 +988,49 @@ def check(session=None):
+        elif dialect.name == 'mysql' and dialect.server_version_info >= (5, 6):
+            conn.execute(text("SELECT GET_LOCK(:id, :timeout)"), id=str(lock), timeout=lock_timeout)
+        elif dialect.name == 'mssql':
+            # TODO: make locking works for MSSQL
```
```suggestion
            # TODO: make locking work for MSSQL
```

uranusjr commented on airflow/utils/db.py:
```diff
@@ -986,3 +988,49 @@ def check(session=None):
+        elif dialect.name == 'mysql' and dialect.server_version_info >= (5, 6):
+            conn.execute(text("select RELEASE_LOCK(:id)"), id=str(lock))
+        elif dialect.name == 'mssql':
+            # TODO: make locking works for MSSQL
```
```suggestion
            # TODO: make locking work for MSSQL
```

uranusjr commented on airflow/utils/db.py:
```diff
@@ -986,3 +988,49 @@ def check(session=None):
+    Cross-db Identifiers for advisory global database locks.
+
+    Postgres uses int64 lock ids so we use the integer value, MySQL uses names, so we use the ``_name_``
+    field.
```
```suggestion
    Postgres uses int64 lock ids so we use the integer value, MySQL uses names, so we
    call ``str()`, which is implemented using the ``_name_`` field.
```

ashb commented on airflow/utils/db.py:
```diff
@@ -986,3 +988,49 @@ def check(session=None):
+    finally:
+        # The session may have been "closed" (which is fine, the lock lasts more than a transaction) -- ensure
+        # we get a usable connection
+        conn = session.connection()
```
This is the problem and the cause of the failures -- if we close the connection we can't unlock someone else's lock.

ashb replied:
> WARNING:  you don't own a lock of type ExclusiveLock

ashb committed:
fixup! Remove magic constants from global DB locks

ashb committed:
Apply suggestions from code review

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

uranusjr reviewed (approved):
MySQL and Postgres tests are all passing, so this should be good to go.
